### PR TITLE
Add keyword to match on 'raw' TLS certificates

### PR DESCRIPTION
**Pull request by thus**
Add new keyword `tls_cert` to match on the 'raw' TLS certificate. This keyword matches on the first certificate in the certificate chain.

Example:
```alert tls any any -> any any (msg:"test tls_cert"; tls_cert; content:"|01 02 03 04|"; sid:1;)```

https://redmine.openinfosecfoundation.org/issues/2670

Prscript:
- PR thus-pcap: https://buildbot.openinfosecfoundation.org/builders/thus-pcap/builds/177
- PR thus: https://buildbot.openinfosecfoundation.org/builders/thus/builds/177

### CODE REVIEW

File: doc/userguide/rules/tls-keywords.rst [hunk 1]
```diff
@@ -116,6 +116,21 @@ Usage::
...
+
+Do a "raw" match on the TLS certificate. This keyword only matches on the
+first certificate in the certificate chain.
+
+Example::
+
+  alert tls any any -> any any (msg:"match bytes in TLS cert"; tls_cert; \
+    content:"|06 09 2a 86|"; sid:200070;)
+
+``tls_cert`` is a 'Sticky buffer'.
```
Comment (norg) [inline doc/userguide/rules/tls-keywords.rst]: **s**ticky instead of **S**ticky? :) but in general doc looks good

File: src/detect-tls-cert.c [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+    DetectBufferSetActiveList(s, g_tls_cert_buffer_id);
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_TLS) != 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *_f,
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: _f is a rust-ism, indicating that the variable isn't used. So seeing it being used as _f is kind of weird :)

File: src/detect-tls-cert.c [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+
+    s = DetectEngineAppendSig(de_ctx, "alert tls any any -> any any "
+                              "(msg:\"Test tls_cert\"; tls_cert; "
+                              "content:\"|06 09 2a 86 48|\"; sid:1;)");
+    FAIL_IF_NULL(s);
+
+    SigGroupBuild(de_ctx);
+    DetectEngineThreadCtxInit(&tv, (void *)de_ctx, (void *)&det_ctx);
+
+    FLOWLOCK_WRLOCK(&f);
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: you can remove the locks from these tests

File: src/detect-tls-cert.c [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+
+    StreamTcpInitConfig(TRUE);
+
+    DetectEngineCtx *de_ctx = DetectEngineCtxInit();
+    FAIL_IF_NULL(de_ctx);
+
+    de_ctx->mpm_matcher = mpm_default_matcher;
+    de_ctx->flags |= DE_QUIET;
+
+    s = DetectEngineAppendSig(de_ctx, "alert tls any any -> any any "
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: declare s here

File: src/detect-tls-cert.c [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+    FLOWLOCK_UNLOCK(&f);
+
+    FAIL_IF(r != 0);
+
+    SigMatchSignatures(&tv, de_ctx, det_ctx, p3);
+
+    FAIL_IF_NOT(PacketAlertCheck(p3, 1));
+    AppLayerParserThreadCtxFree(alp_tctx);
+    DetectEngineThreadCtxDeinit(&tv, det_ctx);
+    SigGroupCleanup(de_ctx);
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: this is called as well by DetectEngineCtxFree

File: src/detect-tls.c [hunk 6]
```diff
@@ -159,9 +159,9 @@ void DetectTlsRegister (void)
     DetectSetupParseRegexes(PARSE_REGEX_FINGERPRINT,
             &fingerprint_parse_regex, &fingerprint_parse_regex_study);
 
-    g_tls_cert_list_id = DetectBufferTypeRegister("tls_cert");
+    g_tls_cert_list_id = DetectBufferTypeRegister("tls_keywords_old");
```
Comment (victorjulien) [inline src/detect-tls.c]: I would suggest tls_legacy_keywords

File: src/detect-tls-cert.c [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+/**
+ * \test Test that a signature containing tls_cert is correctly parsed
+ *       and that the keyword is registred.
+ */
+static int DetectTlsCertTest01(void)
+{
+    DetectEngineCtx *de_ctx = NULL;
+    SigMatch *sm = NULL;
+
+    de_ctx = DetectEngineCtxInit();
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: declare de_ctx here

File: src/detect-tls-cert.c [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+    FAIL_IF_NULL(de_ctx);
+
+    de_ctx->flags |= DE_QUIET;
+    de_ctx->sig_list = SigInit(de_ctx, "alert tls any any -> any any "
+                               "(msg:\"Testing tls_cert\"; tls_cert; "
+                               "content:\"|01 02 03 04 05|\"; sid:1;)");
+    FAIL_IF_NULL(de_ctx->sig_list);
+
+    /* sm should not be in the MATCH list */
+    sm = de_ctx->sig_list->sm_lists[DETECT_SM_LIST_MATCH];
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: declare sm here

File: src/detect-tls-cert.c [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+        }
+
+        InspectionBufferSetup(buffer, cert->cert_data, cert->cert_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+#ifdef UNITTESTS
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: can you move the tests into their own file in tests/ ?

File: src/detect-tls-cert.c [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *_f,
+        const uint8_t _flow_flags, void *txv, const int list_id)
+{
+    BUG_ON(det_ctx->inspect_buffers == NULL);
+    InspectionBuffer *buffer = &det_ctx->inspect_buffers[list_id];
+
+    if (buffer->inspect == NULL) {
+        SSLState *ssl_state = (SSLState *)_f->alstate;
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: can this be const?

File: src/detect-tls-cert.c [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+
+        if (TAILQ_EMPTY(&ssl_state->server_connp.certs)) {
+            return NULL;
+        }
+
+        SSLCertsChain *cert = TAILQ_FIRST(&ssl_state->server_connp.certs);
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: const?

File: src/detect-tls-cert.c [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+ * \param s      Pointer to the Signature to which the keyword belongs
+ * \param str    Should hold an empty string always
+ *
+ * \retval 0 On success
+ * \retval -1 on failure
+ */
+static int DetectTlsCertSetup(DetectEngineCtx *de_ctx, Signature *s,
+                              const char *str)
+{
+    DetectBufferSetActiveList(s, g_tls_cert_buffer_id);
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: check the return code here as well

File: src/detect-tls-cert.c [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+
+/**
+ * \brief Registration function for keyword: tls_cert
+ */
+void DetectTlsCertRegister(void)
+{
+    sigmatch_table[DETECT_AL_TLS_CERT].name = "tls_cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].desc = "content modifier to match the TLS certificate buffer";
+    sigmatch_table[DETECT_AL_TLS_CERT].url = DOC_URL DOC_VERSION "/rules/tls-keywords.html#tls-cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].Match = NULL;
```
Comment (victorjulien) [inline src/detect-tls-cert.c]: remove the NULL settings please, they are redundant

File: doc/userguide/rules/tls-keywords.rst (hunk 1 above)
Comment (thus) [reply doc/userguide/rules/tls-keywords.rst]: It is written as 'Sticky' on all the other keywords in the same file. Do you want me to lowercase the existing keywords as well? :)

File: src/detect-tls-cert.c (hunk 2 above)
Comment (thus) [reply src/detect-tls-cert.c]: I agree! I just copied that from some other code in the repo ;)
It's like that for the other mpm TLS keywords. Do you want me to change this for the other keywords as well?

File: src/detect-tls-cert.c (hunk 3 above)
Comment (thus) [reply src/detect-tls-cert.c]: Sure. Do you want me to remove them from the other mpm TLS keywords too? :)

File: src/detect-tls-cert.c (hunk 9 above)
Comment (thus) [reply src/detect-tls-cert.c]: Sure :)

File: src/detect-tls-cert.c (hunk 4 above)
Comment (thus) [reply src/detect-tls-cert.c]: Will do!

File: src/detect-tls-cert.c (hunk 5 above)
Comment (thus) [reply src/detect-tls-cert.c]: I'll remove it then.

File: src/detect-tls.c (hunk 6 above)
Comment (thus) [reply src/detect-tls.c]: I agree, that's a better name!

File: src/detect-tls-cert.c (hunk 10 above)
Comment (thus) [reply src/detect-tls-cert.c]: I'll change it.

File: src/detect-tls-cert.c (hunk 11 above)
Comment (thus) [reply src/detect-tls-cert.c]: I'll change it.

File: src/detect-tls-cert.c (hunk 12 above)
Comment (thus) [reply src/detect-tls-cert.c]: Will do!

File: src/detect-tls-cert.c (hunk 13 above)
Comment (thus) [reply src/detect-tls-cert.c]: Okay :)

File: src/detect-tls-cert.c (hunk 7 above)
Comment (thus) [reply src/detect-tls-cert.c]: Will do!

File: src/detect-tls-cert.c (hunk 8 above)
Comment (thus) [reply src/detect-tls-cert.c]: Will do!

File: doc/userguide/rules/tls-keywords.rst (hunk 1 above)
Comment (victorjulien) [reply doc/userguide/rules/tls-keywords.rst]: I like the lowercase one better as well. If you want to update it elsewhere, please do.

File: src/detect-tls-cert.c (hunk 2 above)
Comment (victorjulien) [reply src/detect-tls-cert.c]: Yes please.

File: src/detect-tls-cert.c (hunk 3 above)
Comment (victorjulien) [reply src/detect-tls-cert.c]: Yeah I think that would be good.

File: src/detect-tls-cert.c (hunk 12 above)
Comment (victorjulien) [reply src/detect-tls-cert.c]: This is probably something that is missing from a lot of other places too. So maybe you can check at least all the TLS ones for this?

**Comment by thus** [comment]
Replaced by #3851

**Pull request closed without merging**